core/cache/redisBackend.go

```diff
@@ -0,0 +1,378 @@
+		return nil, false
+	}
+
+	b.countHit()
+	return b.buildResult(redisEntry), true
+}
+
+// Set an cache key
+func (b *RedisBackend) Set(key string, entry *Entry) error {
+	conn := b.Pool.Get()
+	defer conn.Close()
+
+	if b.writeLockEnabled {
+		err := b.lock(conn, key)
+		if err != nil {
+			return err
+		}
+	}
+
+	redisEntry := b.buildEntry(entry)
+
+	buffer, err := b.encodeEntry(redisEntry)
+	if err != nil {
+		b.logger.WithField("category", "redisBackend").Error("Error encoding: %v: %v", key, redisEntry)
+		return err
+	}
+
+	err = conn.Send(
+		"SETEX",
+		b.CreatePrefixedKey(key, valuePrefix),
+		int(entry.Meta.gracetime.Sub(time.Now().Round(time.Second))),
+		buffer,
+	)
+	if err != nil {
+		b.logger.WithField("category", "redisBackend").Error("Error setting key %v with timeout %v and buffer %v", key, int(entry.Meta.Gracetime.Seconds()), buffer)
+		return err
+	}
+
+	for _, tag := range entry.Meta.Tags {
+		err = conn.Send(
+			"SADD",
+			b.CreatePrefixedKey(tag, tagPrefix),
+			b.CreatePrefixedKey(key, valuePrefix),
+		)
+		if err != nil {
+			b.logger.WithField("category", "redisBackend").Error("Error setting tag: %v: %v", tag, key)
+			return err
+		}
+	}
+
+	conn.Flush()
+	if b.writeLockEnabled {
+		err = b.unLock(conn, key)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Purge an cache key
+func (b *RedisBackend) Purge(key string) error {
+	conn := b.Pool.Get()
+	defer conn.Close()
+
+	_, err := conn.Do("DEL", b.CreatePrefixedKey(key, valuePrefix))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PurgeTags purges all keys+tags by tag(s)
+func (b *RedisBackend) PurgeTags(tags []string) error {
+	conn := b.Pool.Get()
+	defer conn.Close()
+
+	for _, tag := range tags {
+		reply, err := conn.Do("SMEMBERS", b.CreatePrefixedKey(tag, tagPrefix))
+		members, err := redis.Strings(reply, err)
+		if err != nil {
+			b.logger.WithField("category", "redisBackend").Error(fmt.Sprintf("Failed SMEMBERS for tag '%v': %v", tag, err))
+		}
+
+		for _, member := range members {
+			_, err = conn.Do("DEL", member)
+			if err != nil {
+				b.logger.WithField("category", "redisBackend").Error(fmt.Sprintf("Failed DEL for key '%v': %v", member, err))
+				return err
+			}
+		}
+
+		_, err = conn.Do("DEL", fmt.Sprintf("%v", tag))
+		if err != nil {
+			b.logger.WithField("category", "redisBackend").Error(fmt.Sprintf("Failed DEL for key '%v': %v", tag, err))
+			return err
+		}
+	}
+	conn.Flush()
+
+	return nil
+}
+
+// Flush the whole cache
+func (b *RedisBackend) Flush() error {
+	conn := b.Pool.Get()
+	defer conn.Close()
+
+	err := conn.Send("FLUSHALL")
+	if err != nil {
+		b.logger.WithField("category", "redisBackend").Error(fmt.Sprintf("Failed purge all keys %v", err))
+		return err
+	}
+
+	conn.Flush()
+
+	return nil
+}
+
+func (b *RedisBackend) encodeEntry(entry *RedisCacheEntry) (*bytes.Buffer, error) {
+	gob.Register(entry)
```

Comment:
shouldn't we expect that the caller has called this before?

Reply:
do you mean the "gob.Register" call, or the whole encodeEntry?

Reply:
It's enough to do `func init() { gob.Register(new(RedisCacheEntry)) }` once. gob only needs to look at the struct type, not the values.

Actually, calling gob.Register later, can lead to confusions/race conditions, such as decoding an entry via gob before it was initially registered if using interfaces inside.

Reply:
nice. thanks for the hint!